### PR TITLE
OTP v2

### PR DESCRIPTION
**Pull request by wsp-mbuckley**
Rewrite of the process to use [Open Trip Planner 2.1](https://docs.opentripplanner.org/en/v2.2.0/) and to request individual routes between origin-destination pairs, rather than isochrones. 

- Added generalised cost output matrix
- Added output cost metrics CSV
- Added option to run for various time periods
- Rewritten readme / user guide
- Other fixes / code refactoring

*Note: not all code has been rewritten to TfN's standard's, the focus has mainly been on the new code and areas that were updated for other reasons.*

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,65 +1,277 @@
...
+# OTP4GB-py
 
-This is a port of the [OTP4GB][OTP4GB] tool written by Tom Forth at Open Innovations.
+This tool is a Python wrapper around [Open Trip Planner (OTP) v2](https://docs.opentripplanner.org/en/v2.1.0/)
+which produces cost metrics for public transport routing between given positions in Great Britain.
 
-[OTP4GB]: https://github.com/odileeds/OTP4GB
+This is an updated version of OTP4GB-py produced by Open Innovations (available at
+[github.com/open-innovations/otp4gb-py](https://github.com/open-innovations/otp4gb-py)),
+which itself is a port port of the [OTP4GB](https://github.com/odileeds/OTP4GB)
```
Comment (hulsiejames) [inline README.md:8]: "which itself is a **port port**" --> **port**

File: example-config.yml [hunk 2]
```diff
@@ -1,29 +1,51 @@
...
+    travel_time: 19:00
+    search_window_minutes: 60
+
 modes:
-  - BUS,WALK
-  - TRANSIT,WALK
-  - BICYCLE
-  - CAR
-  - WALK
-centroids: MSOACentroids_NorthEngland_LatLong.csv
```
Comment (hulsiejames) [inline example-config.yml:29]: - centroids: 
^ Parameter is now missing from example-config.yml - has this moved elsewhere?

File: requirements.txt [hunk 3]
```diff
@@ -0,0 +1,17 @@
+python>=3.9
+
+geopandas>=0.11.1
```
Comment (hulsiejames) [inline requirements.txt:3]: Pandas 1.0 [required dependency for GeoPandas](https://geopandas.org/en/stable/getting_started/install.html)? 
(I'm assuming installing geopandas>=0.11.1 will install all required dependencies { inc. pandas 1.0 })

File: setup.py [hunk 4]
```diff
@@ -1,18 +1,67 @@
...
 from urllib.request import urlretrieve
 from subprocess import run
 
 
-MAVEN_PATH=os.path.join('bin', 'apache-maven-3.8.6', 'bin', 'mvn')
+BINARY_FOLDER = pathlib.Path("bin")
+OSM_NAME = "osmconvert64.exe"
+OSMCONVERT_DOWNLOAD_URL = "https://yadi.sk/d/Vnwc4kut3LCBFm"
+# TODO(MB) Find the most up to date version of maven by checking https://dlcdn.apache.org/maven/maven-3
+MAVEN_VERSION = "3.8.7"
```
Comment (hulsiejames) [inline setup.py:13]: This needs to be set as 3.8.8 now. Until L12 TODO is implemented

File: prepare.py [hunk 5]
```diff
@@ -4,139 +4,144 @@
...
-    logger.debug('Extents set to %s', extents)
+    logger.debug("Extents set to %s", extents)
 
     if opt_force:
         shutil.rmtree(filtered_graph_folder, ignore_errors=True)
 
     if os.path.exists(filtered_graph_folder):
         logging.warning(
-            'A folder of filtered GTFS and OSM files already exists. To filter again, delete this folder.')
+            "A folder of filtered GTFS and OSM files already exists. To filter again, delete this folder."
```
Comment (hulsiejames) [inline prepare.py:111]: Or, re-run prepare.py with force overwrite [-F | --force] enabled?

File: README.md [hunk 6]
```diff
@@ -1,65 +1,277 @@
...
+and generalised cost matrix. Before running the process a new folder should be created for storing
+the intemediary files and the outputs, this should also contain the OTP4GB-Py configuration file.
 
-Usage:
+The config file should be named "config.yml" and uses the
+[YAML format](https://docs.ansible.com/ansible/latest/reference_appendices/YAMLSyntax.html#yaml-syntax),
+an [example config](example-config.yml) is provided and the parameters for the it are outlined
+in table 1.
 
+**Table 1:** *Details the parameters available in the config file and their default values.*
```
Comment (hulsiejames) [inline README.md:99]: Newline required between heading & table [for correct rendering](https://stackoverflow.com/a/46697788) - See L181 above Table 2

File: README.md [hunk 7]
```diff
@@ -1,65 +1,277 @@
...
+on the outputs from OTP4GB-Py.
+
+### Accessibility
+The [accessibility](accessibility.py) script uses the zone to zone average durations to calculate
+the amount of various land use metrics which can be reached from a zone in a given time limit. The
+inputs for this script are provided in a config file and outlined in table 3, (see 
+[accessibility config](config/accessibility_config.yml) for an example).
+
+**Table 3:** *Details of the config file parameters for the accessibility script.*
+| Parameter                  | Type               | Default | Description                                                                                                      |
```
Comment (hulsiejames) [inline README.md:258]: Newline required before table

File: process.py [hunk 8]
```diff
@@ -1,119 +1,123 @@
...
 
     # Start OTP Server
     server = Server(opt_base_folder)
-    if not opt_no_server:
-        logger.info('Starting server')
+    if not config.no_server:
+        logger.info("Starting server")
         server.start()
 
     # Load Northern MSOAs
```
Comment (hulsiejames) [inline process.py:56]: Load northern boundaries? (they don't have to be MSOA?)

File: otp4gb/routing.py [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+
+            value = [n for n in match.groups() if n is not None]
+
+        if isinstance(value, (tuple, list)):
+            if len(value) in (2, 3):
+                return geometry.Point(*[float(x) for x in value])
+            raise ValueError(f"point should have 2 coordinates not {len(value)}")
+
+        raise TypeError(
+            f"expected collection or string of coordinates for point not '{value}'"
```
Comment (hulsiejames) [inline otp4gb/routing.py:103]: need `type(value)` rather than `value` ?

**Review by hulsiejames — CHANGES_REQUESTED** [review]
Looks good! One or two changes:

 - within **setup.py** (L13) set maven version to `3.8.8` rather than outdated `3.8.7` until automated checking is implemented
 - within **routing.py** (L103) should `value` not become `type(value)`

______________________________
Some minor formatting within **README.md** 

 - L8 Typo - "which itself is a port port of" to become "a port of"
 - L99 & L258 - Newline often required between tables [for correct rendering](https://stackoverflow.com/a/46697788)

File: example-config.yml (hunk 2 above)
Comment (wsp-mbuckley) [reply example-config.yml:29]: Fixed in 130ec4ba50ac3934ee8259bb4439831a201ebd51

File: setup.py (hunk 4 above)
Comment (wsp-mbuckley) [reply setup.py:13]: Someone fixed within the TRSE branch

File: process.py (hunk 8 above)
Comment (wsp-mbuckley) [reply process.py:56]: Fixed at some point, can't remember when

File: otp4gb/routing.py (hunk 9 above)
Comment (wsp-mbuckley) [reply otp4gb/routing.py:103]: Fixed in 73014cf3fff0ce7ecfc1b0144123aafe469d019c

File: README.md (hunk 7 above)
Comment (wsp-mbuckley) [reply README.md:258]: Fixed in 73014cf3fff0ce7ecfc1b0144123aafe469d019c

File: README.md (hunk 6 above)
Comment (wsp-mbuckley) [reply README.md:99]: Fixed in 73014cf3fff0ce7ecfc1b0144123aafe469d019c

File: prepare.py (hunk 5 above)
Comment (wsp-mbuckley) [reply prepare.py:111]: Fixed in 73014cf3fff0ce7ecfc1b0144123aafe469d019c

File: README.md (hunk 1 above)
Comment (wsp-mbuckley) [reply README.md:8]: Fixed in 73014cf3fff0ce7ecfc1b0144123aafe469d019c

**Comment by wsp-mbuckley** [comment]
Pull request superseded by [OTP_v2_TRSE_](https://github.com/Transport-for-the-North/otp4gb-py/tree/OTP_v2_TRSE_) branch

**Pull request closed without merging**